src/solvers/scip.rs

```diff
@@ -0,0 +1,171 @@
+//! A solver that uses [SCIP](https://scipopt.org), one
+//! of the fastest non-commercial solvers for mixed integer programming.
+
+use std::collections::HashMap;
+
+use russcip::model::Model;
+use russcip::model::ObjSense;
+use russcip::variable::VarType;
+
+use crate::variable::{UnsolvedProblem, VariableDefinition};
+use crate::{
+    constraint::ConstraintReference,
+    solvers::{ObjectiveDirection, ResolutionError, Solution, SolverModel},
+};
+use crate::{Constraint, Variable};
+
+/// The [SCIP](https://scipopt.org) solver,
+/// to be used with [UnsolvedProblem::using].
+pub fn scip(to_solve: UnsolvedProblem) -> SCIPProblem {
+    let mut problem = Model::new();
+
+    problem.include_default_plugins();
+    problem.create_prob("problem");
+    problem.hide_output();
+    problem.set_obj_sense(match to_solve.direction {
+        ObjectiveDirection::Maximisation => ObjSense::Maximize,
+        ObjectiveDirection::Minimisation => ObjSense::Minimize,
+    });
+
+    for (
+        var,
+        &VariableDefinition {
+            min,
+            max,
+            is_integer,
+            ref name,
+        },
+    ) in to_solve.variables.iter_variables_with_def()
+    {
+        let coeff = *to_solve
+            .objective
+            .linear
+            .coefficients
+            .get(&var)
+            .unwrap_or(&0.);
+        let var_type = match is_integer {
+            true => VarType::Integer,
+            false => VarType::Continuous,
+        };
+        problem.add_var(min, max, coeff, name.as_str().clone(), var_type);
+    }
+
+    SCIPProblem { problem }
+}
+
+/// A SCIP Model
+pub struct SCIPProblem {
+    problem: Model,
+}
+
+impl SolverModel for SCIPProblem {
+    type Solution = SCIPSolution;
+    type Error = ResolutionError;
+
+    fn solve(self) -> Result<Self::Solution, Self::Error> {
+        let vars = self.problem.get_vars();
+        self.problem.solve();
+        let sol = self.problem.get_best_sol();
+        let values = vars
+            .iter()
+            .map(|var| (var.get_index(), sol.get_var_val(var)))
+            .collect();
```

Comment:
Maybe we could collect the values only in the OPTIMAL case ?

Reply:
Done.